[SPARK-13947][SQL] The error message from using an invalid column reference is not clear

rberenguel opened the pull request:
## What changes were proposed in this pull request?

 Rewritten error message for clarity. Added extra information in case of attribute name collision, hinting the user to double-check referencing two different tables

## How was this patch tested?

No functional changes, only final message has changed. It has been tested manually against the situation proposed in the JIRA ticket. Automated tests in repository pass.

This PR is original work from me and I license this work to the Spark project

rberenguel committed:
SPARK-13947 Rewritten error message for clarity. Added extra information in case of attribute name collision

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins, ok to test.

holdenk commented:
@rberenguel : how about adding the "[SQL]" tag to this, since while the feature request comes out of PySpark its changing the SQL code.

SparkQA commented:
**[Test build #73604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73604/testReport)** for PR 17100 at commit [`981224b`](https://github.com/apache/spark/commit/981224bd9631a0ac633b6967e639197794894fcf).

SparkQA commented:
**[Test build #73604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73604/testReport)** for PR 17100 at commit [`981224b`](https://github.com/apache/spark/commit/981224bd9631a0ac633b6967e639197794894fcf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73604/
Test FAILed.

rberenguel commented:
@holdenk added! And I suspect I have some issue locally with coursier when running the tests... They supposedly work (as in, no outputted errors/failures?) 1/3 times, the other 2 cause coursier cache misses. But I think the no errors one is actually "I give up running the test suite anymore". I'll give a look at this (since it's pretty bad I can't rely on the local test suite!) and fix the failing test, is pretty straightforward.

rberenguel committed:
SPARK-13947 Regenerated the golden file with the updated error message. Added a mental note to disable coursier when running the full Spark test suite

SparkQA commented:
**[Test build #73700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73700/testReport)** for PR 17100 at commit [`ea07688`](https://github.com/apache/spark/commit/ea07688ae1e09505fef58de834c35e08324cb83b).

SparkQA commented:
**[Test build #73700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73700/testReport)** for PR 17100 at commit [`ea07688`](https://github.com/apache/spark/commit/ea07688ae1e09505fef58de834c35e08324cb83b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73700/
Test FAILed.

rberenguel committed:
SPARK-13947 Regenerated golden file after formatting fix. Fixed the AnalysisErrorSuite test that was failing. Just in case, removed the hardcoding of the names and hashes from this test

SparkQA commented:
**[Test build #73714 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73714/testReport)** for PR 17100 at commit [`65b9596`](https://github.com/apache/spark/commit/65b9596c229ac2b62ecdfeb98e541d2ea92e078d).

SparkQA commented:
**[Test build #73714 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73714/testReport)** for PR 17100 at commit [`65b9596`](https://github.com/apache/spark/commit/65b9596c229ac2b62ecdfeb98e541d2ea92e078d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73714/
Test PASSed.

holdenk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -338,11 +338,29 @@ trait CheckAnalysis extends PredicateHelper {
             failAnalysis(
-              s"resolved attribute(s) $missingAttributes missing from $input " +
-                s"in operator ${operator.simpleString}")
+              s"|Some resolved attribute(s) are not present among available attributes " +
```
This is really long, would maybe """ + stripMargin be easier to write/read?

holdenk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -398,16 +398,22 @@ class AnalysisErrorSuite extends AnalysisTest {
 
-    assertAnalysisError(plan, "resolved attribute(s) a#1L missing from a#2L" :: Nil)
+    assertAnalysisError(plan,
+                        "Some resolved attribute(s) are not present among available " +
```
This is a weird mixing of strings formats.

holdenk reviewed (commented):
Thanks for working on this, maybe we can ask @marmbrus to take a look at this since its mostly a SQL change.

Commit:
SPARK-13947 Code review improvements

Commit:
Merge branch 'master' into SPARK-13947-error-message

# Conflicts:
#	sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out

SparkQA commented:
**[Test build #75013 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75013/testReport)** for PR 17100 at commit [`7bb6f35`](https://github.com/apache/spark/commit/7bb6f35d9153ea522b8ccb4bcd44bf1c70cc64f0).

SparkQA commented:
**[Test build #75013 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75013/testReport)** for PR 17100 at commit [`7bb6f35`](https://github.com/apache/spark/commit/7bb6f35d9153ea522b8ccb4bcd44bf1c70cc64f0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75013/
Test FAILed.

gatorsmile commented:
Can you remove `[PYTHON]` from the PR title? Thanks!

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -337,11 +337,29 @@ trait CheckAnalysis extends PredicateHelper {
+
+            val repeatedNameHint = if (commonNames.size > 0) {
+              s"\n|Observe that attribute(s) ${commonNames.mkString(",")} appear in your " +
+                "query with at least two different hashes, but same name."
```
Users do not understand what hashes mean.

rberenguel replied:
Good point!

gatorsmile commented:
To detect duplicate names, you need to pass SQLConf `conf` in `checkAnalysis` for supporting the case sensitivity in name resolution. For each missing attribute, you can try using `conf.resolver` to find whether there exists a column from `o.inputSet` whose name is the same. After detecting the duplicate names, you can output the message to say the missing attributes are from a Dataset different from the input set.

rberenguel commented:
@gatorsmile Thanks for the pointers, finally found some time to come back to this. I'm not sure if my approach to get the `SQLConf` into `checkAnalysis` is the correct one in my current local changes (since it seems to change a possible API endpoint). I changed the current implementation in the trait to be named instead `def checkAnalysisWithConf(plan: LogicalPlan, conf: SQLConf): Unit` and added an abstract method `def checkAnalysis(plan: LogicalPlan): Unit` that is then implemented in `Analyzer` (where we have a `conf` we can pass around). I haven't fixed all the rest yet, was puzzled enough with the correctness of this for now ;) Thanks

Commit:
SPARK-13947 Not sure about these changes, but passes the local suite just fine. Let’s see!

SparkQA commented:
**[Test build #76401 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76401/testReport)** for PR 17100 at commit [`4ac8143`](https://github.com/apache/spark/commit/4ac8143cf63f5b4777a66f236824671b0bb05933).

rberenguel committed:
Merge branch 'master' into SPARK-13947-error-message

SparkQA commented:
**[Test build #76402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76402/testReport)** for PR 17100 at commit [`766a033`](https://github.com/apache/spark/commit/766a033f19602dbd7da6eff96947236c8c0fd2a2).

SparkQA commented:
**[Test build #76402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76402/testReport)** for PR 17100 at commit [`766a033`](https://github.com/apache/spark/commit/766a033f19602dbd7da6eff96947236c8c0fd2a2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76402/
Test FAILed.

SparkQA commented:
**[Test build #76401 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76401/testReport)** for PR 17100 at commit [`4ac8143`](https://github.com/apache/spark/commit/4ac8143cf63f5b4777a66f236824671b0bb05933).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76401/
Test PASSed.

Commit:
SPARK-13947 Never quick-edit a non-clean merge after 1 AM

SparkQA commented:
**[Test build #76434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76434/testReport)** for PR 17100 at commit [`c2dfe11`](https://github.com/apache/spark/commit/c2dfe11c4c5da3ec8fb0983377400aa999213c35).

SparkQA commented:
**[Test build #76434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76434/testReport)** for PR 17100 at commit [`c2dfe11`](https://github.com/apache/spark/commit/c2dfe11c4c5da3ec8fb0983377400aa999213c35).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76434/
Test PASSed.

rberenguel commented:
Can you give a look to the changes @gatorsmile when you have a few spare moments? Also not sure how the conflict has appeared: it was merging cleanly when the CI test suite ran (will obviously fix, but first I want to confirm it's the right way of doing this now)

rberenguel committed:
Merge branch 'master' into SPARK-13947-error-message

rberenguel committed:
Merge branch 'master' into SPARK-13947-error-message

SparkQA commented:
**[Test build #77502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77502/testReport)** for PR 17100 at commit [`0cb9825`](https://github.com/apache/spark/commit/0cb9825a1cfebe67f08ce344e67a06c53e4a78be).

SparkQA commented:
**[Test build #77502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77502/testReport)** for PR 17100 at commit [`0cb9825`](https://github.com/apache/spark/commit/0cb9825a1cfebe67f08ce344e67a06c53e4a78be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77502/
Test PASSed.

holdenk commented:
re-ping @gatorsmile?

gatorsmile commented:
Will review it this weekend. Thanks!

rberenguel committed:
Merge branch 'master' into SPARK-13947-error-message

SparkQA commented:
**[Test build #82950 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82950/testReport)** for PR 17100 at commit [`6bbe496`](https://github.com/apache/spark/commit/6bbe496f9d8bf48953ba15518d5f80f0b4140ed5).

SparkQA commented:
**[Test build #82950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82950/testReport)** for PR 17100 at commit [`6bbe496`](https://github.com/apache/spark/commit/6bbe496f9d8bf48953ba15518d5f80f0b4140ed5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82950/
Test FAILed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +273,25 @@ trait CheckAnalysis extends PredicateHelper {
 
         operator match {
           case o if o.children.nonEmpty && o.missingInput.nonEmpty =>
+            val resolver = conf.resolver
```
We do not need to pass conf now. You can directly call `plan.conf.resolver`

rberenguel replied:
Oh, much handier. Thanks!

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -125,7 +128,7 @@ trait CheckAnalysis extends PredicateHelper {
 
           case s: SubqueryExpression =>
-            checkSubqueryExpression(operator, s)
+            checkAnalysisWithConf(s.plan, conf)
```
This is wrong. Could you revert it back?

rberenguel replied:
Yup, sorry. Was too eager to merge back master

gatorsmile commented:
cc @wzhfy Could you review this PR?

viirya commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -408,16 +408,28 @@ class AnalysisErrorSuite extends AnalysisTest {
-    assertAnalysisError(plan, "resolved attribute(s) a#1L missing from a#2L" :: Nil)
+    val errorMsg = s"""Some resolved attribute(s) are not present among the available attributes
+                     |for a query.
+                     |${aId.mkString("#")}L is not in ${otherAId.mkString("#")}L.
```
Isn't `${aId.mkString("#")}L` just `attrA.toString`?

rberenguel replied:
Yes thanks, I discovered this recently, but didn't come back to this PR. Will fix

viirya replied:
Seems this is still unchanged?

wzhfy replied:
Can you `{aId.mkString("#")}L` to `attrA.toString`? and also remove `val aId` and `val otherAId`.

rberenguel replied:
Oh, I did it wrongly and thus did no work (did aId.toString and got a horrible java String object view). Changing it, also regenerating the golden files which seemed to make the test fail

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +273,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val availableAttributes = o.inputSet.mkString(",")
+            val repeatedNameHint = if (commonAttrs.size > 0) {
+              val commonNames = commonAttrs.map(_.name).mkString(",")
+              s"""\n|Observe that attribute(s) $commonNames appear in two
```
I'm not sure if this is always the case. Some failures in the query plan analysis can cause missing attributes, I mean.

viirya replied:
The new line `\n` is intentional?

rberenguel replied:
I'll rewrite the wording a bit so it hints to this as the cause (it's the most usual I suspect). The newline is for readability of the error message. I could remove it.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +273,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val repeatedNameHint = if (commonAttrs.size > 0) {
+              val commonNames = commonAttrs.map(_.name).mkString(",")
+              s"""\n|Observe that attribute(s) $commonNames appear in two
+                  |different datasets, with the same name"""
```
Missing a `.` at the end of sentence.

rberenguel replied:
Thanks!

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +273,25 @@ trait CheckAnalysis extends PredicateHelper {
         operator match {
           case o if o.children.nonEmpty && o.missingInput.nonEmpty =>
+            val resolver = conf.resolver
+            val commonAttrs = o.missingInput.filter(x =>
```
Not quite common attrs actually. Maybe `attrsWithSameName`.

rberenguel replied:
Indeed!

viirya commented:
In the title, why specifying "PySpark" DataFrames? I don't see any python code change.

rberenguel commented:
@viirya Hi, I based the title of the PR off the original ticket (https://issues.apache.org/jira/browse/SPARK-13947), but indeed has nothing to do with Python. I see now the ticket has been "moved" to SQL, so I'll rewrite the title when I send the fixes (might take me a while to re-run the test suite)

rberenguel committed:
SPARK-13947 Code review modifications. AnalysisErrorSuite passes, but I get out of memory locally for a full suite

SparkQA commented:
**[Test build #82960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82960/testReport)** for PR 17100 at commit [`75be390`](https://github.com/apache/spark/commit/75be3900f65e64bd083a91e5c63978489a61ccfe).

SparkQA commented:
**[Test build #82960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82960/testReport)** for PR 17100 at commit [`75be390`](https://github.com/apache/spark/commit/75be3900f65e64bd083a91e5c63978489a61ccfe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rberenguel commented:
Applied all changes @viirya & @gatorsmile

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82960/
Test PASSed.

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -24,6 +24,7 @@ import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
 import org.apache.spark.sql.catalyst.optimizer.BooleanSimplification
 import org.apache.spark.sql.catalyst.plans._
 import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.internal.SQLConf
```
unused import

rberenguel replied:
Thanks

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -24,6 +24,7 @@ import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
 import org.apache.spark.sql.catalyst.optimizer.BooleanSimplification
 import org.apache.spark.sql.catalyst.plans._
 import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.internal.SQLConf
```
No need to import SQLConf now.

rberenguel replied:
Thanks

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +271,25 @@ trait CheckAnalysis extends PredicateHelper {
-                s"in operator ${operator.simpleString}")
+              s"""Some resolved attribute(s) are not present among the available attributes
+                |for a query.
+                |$missingAttributes is not in $availableAttributes.$repeatedNameHint
```
Move `$repeatedNameHint` in next line as:
```scala
|$missingAttributes is not in $availableAttributes.
|$repeatedNameHint
```
So you can remove the new line above?

rberenguel replied:
Oh, thanks. I didn't see it

wzhfy replied:
This may add an empty line if `repeatedNameHint` is an empty string?

rberenguel replied:
Dang, I knew there was a reason for the weird new line back when I wrote this. I'm putting the `\n` back, do you have any idea that doesn't look as weird as this does, though?

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +271,25 @@ trait CheckAnalysis extends PredicateHelper {
           case o if o.children.nonEmpty && o.missingInput.nonEmpty =>
+            val resolver = plan.conf.resolver
+            val attrsWithSameName = o.missingInput.filter(x =>
+              o.inputSet.exists(y => resolver(x.name, y.name)))
```
nit:
```
val attrsWithSameName = o.missingInput.filter { missing =>
  o.inputSet.exists(input => resolver(missing.name, input.name))
}
```

rberenguel replied:
Thanks

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +271,25 @@ trait CheckAnalysis extends PredicateHelper {
             val missingAttributes = o.missingInput.mkString(",")
-            val input = o.inputSet.mkString(",")
+            val availableAttributes = o.inputSet.mkString(",")
+            val repeatedNameHint = if (attrsWithSameName.size > 0) {
```
`missingAttributes` doesn't need `missingAttributes` or `availableAttributes`, could you move this definition above right after `attrsWithSameName`?

wzhfy replied:
`attrsWithSameName.size > 0` -> `attrsWithSameName.nonEmpty`

rberenguel replied:
I don't understand where I should move missing or available, isn't it already just right after `attrsWithSameName`? But I moved it after repeatedNameHint, since it's needed only for the `failAnalysis` argument

wzhfy replied:
Yea that' what I meant :)

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +271,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val repeatedNameHint = if (attrsWithSameName.size > 0) {
+              val commonNames = attrsWithSameName.map(_.name).mkString(",")
+              s"""\n|Attribute(s) `$commonNames` seem to appear in two
+                  |different datasets, with the same name."""
```
The `datasets` concept is a little strange here, would `inputs` or `input operators` be better?

wzhfy replied:
`\n|Please check attribute(s) `$commonNames`, they seem to appear in two...`

rberenguel replied:
I have changed it. Agree datasets may not be the best word to define it

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +271,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val repeatedNameHint = if (attrsWithSameName.size > 0) {
+              val commonNames = attrsWithSameName.map(_.name).mkString(",")
+              s"""\n|Attribute(s) `$commonNames` seem to appear in two
+                  |different datasets, with the same name."""
```
I'm +0 for this additional error msg. I'm not sure if it is actually more clear to end users.

rberenguel replied:
What do you suggest instead?

viirya replied:
Original error msg looks good to me, actually. But this change seems ok to me.

rberenguel committed:
SPARK-13947 Another batch of code review modifications

SparkQA commented:
**[Test build #82970 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82970/testReport)** for PR 17100 at commit [`6e8ab42`](https://github.com/apache/spark/commit/6e8ab429a229df5cb075239c69fe90404734ef64).

SparkQA commented:
**[Test build #82970 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82970/testReport)** for PR 17100 at commit [`6e8ab42`](https://github.com/apache/spark/commit/6e8ab429a229df5cb075239c69fe90404734ef64).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82970/
Test FAILed.

wzhfy commented:
@rberenguel Could you fix [this comment](https://github.com/apache/spark/pull/17100#pullrequestreview-71023434)?

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,27 @@ trait CheckAnalysis extends PredicateHelper {
+                |$missingAttributes is not in $availableAttributes.
+                |$repeatedNameHint
+                |The failed query was for operator
+                |${operator.simpleString}""".stripMargin)
```
Actually I agree with @viirya, the original error message is OK for me. If you really want to improve it, I think just adding `repeatedNameHint` at the end would be good enough:
```
s"Resolved attribute(s) $missingAttributes missing from $input in operator ${operator.simpleString}. $repeatedNameHint"
```
What do you think?

rberenguel replied:
This is good for me, having the `repeatedNameHint` is the whole point, at the end.

wzhfy replied:
the final message could be:
```
if (repeatedNameHint.nonEmpty) msg + “ ” + repeatedNameHint else msg
```

rberenguel replied:
Hmmm seems reasonable, indeed. and avoids that horrible trailing newline. Thanks a lot :)

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,27 @@ trait CheckAnalysis extends PredicateHelper {
           case o if o.children.nonEmpty && o.missingInput.nonEmpty =>
+            val resolver = plan.conf.resolver
+            val attrsWithSameName = o.missingInput.filter(missing =>
+              o.inputSet.exists(input => resolver(missing.name, input.name)))
```
could you change the format?
```
o.missingInput.filter { missing =>
  o.inputSet.exists(input => resolver(missing.name, input.name))
}
```

rberenguel replied:
Done. Do you know if there is any style rule for when to use parentheses vs braces in these kind of lambda functions?

wzhfy replied:
Usually parentheses are used in one line, while curly braces are used in multi-line. For complete scala style guide, please check this: https://github.com/databricks/scala-style-guide

rberenguel replied:
Thanks @wzhfy

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,27 @@ trait CheckAnalysis extends PredicateHelper {
+
             val missingAttributes = o.missingInput.mkString(",")
-            val input = o.inputSet.mkString(",")
+            val availableAttributes = o.inputSet.mkString(",")
```
could you revert this change? original `input` is ok for me.

rberenguel replied:
Okies

rberenguel committed:
SPARK-13947 Another batch of code review modifications

rberenguel committed:
SPARK-13947 Format change for a lambda

rberenguel committed:
SPARK-13947 Improve the final formatting

SparkQA commented:
**[Test build #82980 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82980/testReport)** for PR 17100 at commit [`33a8a71`](https://github.com/apache/spark/commit/33a8a71092b699f6e402c67684fe4bfab6d2ccf8).

SparkQA commented:
**[Test build #82979 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82979/testReport)** for PR 17100 at commit [`72b72cf`](https://github.com/apache/spark/commit/72b72cfb0fecc169e0bc289f7f319bdb2b947de5).

SparkQA commented:
**[Test build #82981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82981/testReport)** for PR 17100 at commit [`e3cc455`](https://github.com/apache/spark/commit/e3cc4555b7a14299b93e8e4e0ae6d7f482582f9d).

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,25 @@ trait CheckAnalysis extends PredicateHelper {
+              o.inputSet.exists(input => resolver(missing.name, input.name))
+            }
+            val repeatedNameHint = if (attrsWithSameName.nonEmpty) {
+              val commonNames = attrsWithSameName.map(_.name).mkString(",")
```
`commonNames` -> `sameNames`?

rberenguel replied:
Done

wzhfy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -408,16 +408,23 @@ class AnalysisErrorSuite extends AnalysisTest {
+    val attrA = AttributeReference("a", LongType)(exprId = ExprId(1))
+    val otherA = AttributeReference("a", LongType)(exprId = ExprId(2))
+    val bAlias = Alias(sum(attrA), "b")() :: Nil
+    val plan = Aggregate(
```
we can make the test case more strong by adding another attribute `c` which is missing from input but doesn't have the same name in input.

rberenguel replied:
Added another attribute

wzhfy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -408,16 +408,23 @@ class AnalysisErrorSuite extends AnalysisTest {
+                     |Please check attribute(s) `a`, they seem to appear in two
+                     |different input operators, with the same name.""".stripMargin
+
+
```
only one new line is enough here

rberenguel replied:
Fixed

SparkQA commented:
**[Test build #82981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82981/testReport)** for PR 17100 at commit [`e3cc455`](https://github.com/apache/spark/commit/e3cc4555b7a14299b93e8e4e0ae6d7f482582f9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82981/
Test PASSed.

SparkQA commented:
**[Test build #82979 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82979/testReport)** for PR 17100 at commit [`72b72cf`](https://github.com/apache/spark/commit/72b72cfb0fecc169e0bc289f7f319bdb2b947de5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82979/
Test PASSed.

SparkQA commented:
**[Test build #82980 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82980/testReport)** for PR 17100 at commit [`33a8a71`](https://github.com/apache/spark/commit/33a8a71092b699f6e402c67684fe4bfab6d2ccf8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82980/
Test PASSed.

rberenguel committed:
SPARK-13947 Test and message improvements

SparkQA commented:
**[Test build #82994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82994/testReport)** for PR 17100 at commit [`34753b5`](https://github.com/apache/spark/commit/34753b513323f5076edd4c5006983a9c9d3d97d7).

SparkQA commented:
**[Test build #82994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82994/testReport)** for PR 17100 at commit [`34753b5`](https://github.com/apache/spark/commit/34753b513323f5076edd4c5006983a9c9d3d97d7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82994/
Test PASSed.

rberenguel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val repeatedNameHint = if (attrsWithSameName.nonEmpty) {
+              val commonNames = attrsWithSameName.map(_.name).mkString(",")
+              s"""|Please check attribute(s) `$commonNames`, they seem to appear in two
+                  |different input operators, with the same name.""".stripMargin
```
I have changed it, now looks better no matter how many appear

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val repeatedNameHint = if (attrsWithSameName.nonEmpty) {
+              val sameNames = attrsWithSameName.map(_.name).mkString(",")
+              s"""Attribute(s) with the same name appear in the operation: `$sameNames`.
+                  |Please check if the right attribute(s) are used.""".stripMargin
```
Personally I prefer one line message because it's an parameter of `AnalysisException`, but that's not a strong point.

rberenguel replied:
I don't have a strong case for having two lines here, and I see the point of seeing it as one line. I guess the best way to have it as one line is splitting it into 
```scala
val attributeRepetitionMsg = s"Attribute(s) with the same name appear in the operation: `$sameNames`"
val checkAttributesMsg = s"Please check if the right attribute(s) are used."
```
to avoid hitting the < 100 chars linting rule?

wzhfy commented:
LGTM except one minor comment. ping @gatorsmile

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,25 @@ trait CheckAnalysis extends PredicateHelper {
+            }
+            val repeatedNameHint = if (attrsWithSameName.nonEmpty) {
+              val sameNames = attrsWithSameName.map(_.name).mkString(",")
+              s"""Attribute(s) with the same name appear in the operation: `$sameNames`.
```
Normally, we do not quote multiple column names in the same quote.

rberenguel replied:
You mean wrapping each column name with the ticks?

gatorsmile replied:
Let us follow what we did for `missingAttributes`. No need to do it.

rberenguel replied:
Ok

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -270,12 +270,25 @@ trait CheckAnalysis extends PredicateHelper {
+            val msg = s"""Resolved attribute(s) $missingAttributes missing from $input
+                          |in operator ${operator.simpleString}.""".stripMargin
+
+            failAnalysis(if (repeatedNameHint.nonEmpty) msg + "\n" + repeatedNameHint else msg)
```
How about 
```
            val missingAttributes = o.missingInput.mkString(",")
            val input = o.inputSet.mkString(",")
            val msgForMissingAttributes = s"Resolved attribute(s) $missingAttributes missing " +
              s"from $input in operator ${operator.simpleString}."

            val resolver = plan.conf.resolver
            val attrsWithSameName = o.missingInput.filter { missing =>
              o.inputSet.exists(input => resolver(missing.name, input.name))
            }

            val msg = if (attrsWithSameName.nonEmpty) {
              val sameNames = attrsWithSameName.map(_.name).mkString(",")
              s"$msgForMissingAttributes. Attribute(s) with the same name appear in the " +
                s"operation: $sameNames. Please check if the right attribute(s) are used."
            } else {
              msgForMissingAttributes
            }

            failAnalysis(msg)
```

rberenguel replied:
Yup thanks, is more orderly this way

rberenguel committed:
SPARK-13947 More code review changes

SparkQA commented:
**[Test build #83020 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83020/testReport)** for PR 17100 at commit [`b8fbdea`](https://github.com/apache/spark/commit/b8fbdea3b35ea69e6dc74ea9c11e257ac88dc765).

SparkQA commented:
**[Test build #83020 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83020/testReport)** for PR 17100 at commit [`b8fbdea`](https://github.com/apache/spark/commit/b8fbdea3b35ea69e6dc74ea9c11e257ac88dc765).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83020/
Test PASSed.

gatorsmile commented:
Thanks! Merged to master.

rberenguel commented:
Thanks @gatorsmile @holdenk @viirya @wzhfy for all the help

holdenk commented:
Congratulations on your change @rberenguel :D :)

gatorsmile commented:
@rberenguel Thank you for your contribution!
